db/query_test.go

```diff
@@ -568,3 +568,98 @@ func countQueryResults(results sgbucket.QueryResultIterator) int {
 	}
 	return count
 }
+
+func TestQueryChannelsActiveOnlyWithLimit(t *testing.T) {
+
+	if base.UnitTestUrlIsWalrus() {
+		t.Skip("This test require Couchbase Server")
+	}
+
+	db, testBucket := setupTestDB(t)
+	defer testBucket.Close()
+	defer tearDownTestDB(t, db)
+
+	// Create doc1, revision 1-a
+	body := Body{"channels": []string{"ABC"}, "name": "Alice"}
+	doc, _, err := db.PutExistingRevWithBody("doc1", body, []string{"1-a"}, false)
+	require.NoError(t, err, "Couldn't create document")
+	startSeq := doc.Sequence
+
+	// Tombstone doc1, revision 2-a
+	body[BodyDeleted] = true
+	doc, _, err = db.PutExistingRevWithBody("doc1", body, []string{"2-a", "1-a"}, false)
+	require.NoError(t, err, "Couldn't create document")
+
+	// Create doc2 and two conflicting changes:
+	body["name"] = "Bob"
+	_, _, err = db.PutExistingRevWithBody("doc2", body, []string{"1-a"}, false)
+	require.NoError(t, err, "Couldn't create document")
+
+	body["name"] = "Owen"
+	_, _, err = db.PutExistingRevWithBody("doc2", body, []string{"2-b", "1-a"}, false)
+	require.NoError(t, err, "Couldn't create revision 2-b of doc2")
+
+	body["name"] = "Chloe"
+	_, _, err = db.PutExistingRevWithBody("doc2", body, []string{"2-a", "1-a"}, false)
+	require.NoError(t, err, "Couldn't create revision 2-a of doc2")
+
+	// Tombstone the non-winning branch of a conflicted document
+	body[BodyDeleted] = true
+	_, _, err = db.PutExistingRevWithBody("doc2", body, []string{"3-a", "2-a"}, false)
+	assert.NoError(t, err, "Add 3-a (tombstone)")
+
+	// Create doc3
+	body["name"] = "Emily"
+	doc, _, err = db.PutExistingRevWithBody("doc3", body, []string{"1-a"}, false)
+	require.NoError(t, err, "Couldn't create revision 2-a of doc3")
+	endSeq := doc.Sequence
+
+	// Issue channel query with limit and activeOnly true
+	entries, err := db.getChangesInChannelFromQuery("ABC", startSeq, endSeq, 2, true)
+	assert.Len(t, entries, 2)
+	assert.Equal(t, "doc2", entries[0].DocID)
+	assert.Equal(t, "2-b", entries[0].RevID)
+	assert.Equal(t, "doc3", entries[1].DocID)
+	assert.Equal(t, "1-a", entries[1].RevID)
+	assert.Equal(t, uint8(0x14), entries[0].Flags)
+	assert.Equal(t, uint8(0x0), entries[1].Flags)
+
+	// Issue channel query with no limit and activeOnly true
+	entries, err = db.getChangesInChannelFromQuery("ABC", startSeq, endSeq, 0, true)
+	assert.Len(t, entries, 3)
+	assert.Equal(t, "doc1", entries[0].DocID)
+	assert.Equal(t, "2-a", entries[0].RevID)
+	assert.Equal(t, "doc2", entries[1].DocID)
+	assert.Equal(t, "2-b", entries[1].RevID)
+	assert.Equal(t, "doc3", entries[2].DocID)
+	assert.Equal(t, "1-a", entries[2].RevID)
+	assert.Equal(t, uint8(0x1), entries[0].Flags)
+	assert.Equal(t, uint8(0x14), entries[1].Flags)
+	assert.Equal(t, uint8(0x0), entries[2].Flags)
+
+	// Issue channel query with limit and activeOnly false
+	entries, err = db.getChangesInChannelFromQuery("ABC", startSeq, endSeq, 3, false)
+	assert.Len(t, entries, 3)
```

Comment:
Testing with a limit value that's equal to the total number of entries in the database doesn't verify whether limit is working.

Reply:
Modified the limit to 2 which is less than the total number of entries in the database to make sure limit is working.